runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+                    success_value: None,
+                    error: Some("Invalid target: target is address corresponding to existing named account_id".into()),
+                });
+            }
+        } else {
+            let ext = WalletContract::ext(current_account_id).with_unused_gas_weight(1);
+            match action_to_promise(target, action).map(|p| p.then(ext.rlp_execute_callback())) {
+                Ok(p) => p,
+                Err(e) => {
+                    return PromiseOrValue::Value(e.into());
+                }
+            }
+        };
+        PromiseOrValue::Promise(promise)
+    }
+
+    #[private]
+    pub fn rlp_execute_callback(&mut self) -> ExecuteResponse {
+        let n = env::promise_results_count();
+        let mut success_value = None;
+        for i in 0..n {
+            match env::promise_result(i) {
+                PromiseResult::Failed => {
+                    return ExecuteResponse {
+                        success: false,
+                        success_value: None,
+                        error: Some("Failed Near promise".into()),
+                    };
+                }
+                PromiseResult::Successful(value) => success_value = Some(value),
+            }
+        }
+        ExecuteResponse {
+            success: true,
+            success_value,
+            error: None,
+        }
+    }
+
+    #[private]
+    pub fn ban_relayer(&mut self) -> ExecuteResponse {
+        ExecuteResponse {
+            success: false,
+            success_value: None,
+            error: Some("Error: faulty relayer".into()),
+        }
+    }
+}
+
+fn inner_rlp_execute(
+    current_account_id: AccountId,
+    predecessor_account_id: AccountId,
+    target: AccountId,
+    tx_bytes_b64: String,
+    nonce: &mut u64,
+) -> Result<Promise, Error> {
+    if *nonce == u64::MAX {
+        return Err(Error::AccountNonceExhausted);
+    }
+    let context = ExecutionContext::new(
+        current_account_id.clone(),
+        predecessor_account_id,
+        env::attached_deposit(),
+    )?;
+
+    let (action, validation_outcome) =
+        internal::parse_rlp_tx_to_action(&tx_bytes_b64, &target, &context, nonce)?;
+    let promise = match validation_outcome {
+        TransactionValidationOutcome::Validated => {
+            let ext = WalletContract::ext(current_account_id).with_unused_gas_weight(1);
+            action_to_promise(target, action)?.then(ext.rlp_execute_callback())
+        }
+        TransactionValidationOutcome::AddressCheckRequired(address) => {
+            let ext = WalletContract::ext(current_account_id).with_unused_gas_weight(1);
+            let address_registrar = {
+                let account_id = ADDRESS_REGISTRAR_ACCOUNT_ID
+                    .trim()
+                    .parse()
+                    .unwrap_or_else(|_| env::panic_str("Invalid address registrar"));
+                ext_registrar::ext(account_id).with_static_gas(Gas::from_tgas(5))
```

Comment:
Why 5 here?

Reply:
The call to `lookup` in the registrar contract is very simple, just one storage read. So 5 Tgas should be enough for it.